// waku/dist/lib/config.js
const DEFAULT_HTML_HEAD = `
<meta charset="utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
`.trim();
const ADDITIONAL_HTML_HEAD = `
<meta name="generator" content="Waku" />
`.trim();
const DO_NOT_BUNDLE = '';
const DEFAULT_MIDDLEWARE = (cmd)=>[
        ...cmd === 'dev' ? [
            import(/* @vite-ignore */ DO_NOT_BUNDLE + 'waku/middleware/dev-server')
        ] : [],
        import('waku/middleware/ssr'),
        import('waku/middleware/rsc')
    ];
// Keep async function for future extension
export async function resolveConfig(config) {
    const resolvedConfig = {
        basePath: '/',
        srcDir: 'src',
        distDir: 'dist',
        preserveModuleDirs: [
            'pages',
            'templates',
            'routes',
            'components'
        ],
        privateDir: 'private',
        rscPath: 'RSC',
        htmlAttrs: '',
        htmlHead: DEFAULT_HTML_HEAD,
        middleware: DEFAULT_MIDDLEWARE,
        ...config
    };
    if (!resolvedConfig.htmlHead.includes(ADDITIONAL_HTML_HEAD)) {
        resolvedConfig.htmlHead += ADDITIONAL_HTML_HEAD;
    }
    return resolvedConfig;
}
export const EXTENSIONS = [
    '.js',
    '.ts',
    '.tsx',
    '.jsx',
    '.mjs',
    '.cjs'
];
